runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -6567,12 +6567,11 @@ static void genInitObjectHeader(TR::Node             *node,
                            (!TR::Compiler->om.generateCompressedObjectHeaders() ||
                             (cg->comp()->compileRelocatableCode() && cg->comp()->getOption(TR_UseSymbolValidationManager)));
 
-   // This code was moved to this point so that the romClass can be used by the AOT
-   // portion without calling the method again.
-   //
-   J9ROMClass *romClass = 0;
-   TR_ASSERT(clazz, "Cannot have a null OpaqueClassBlock\n");
-   romClass = TR::Compiler->cls.romClassOf(clazz);
+   // Dynamic array cloning is true when the class (2nd) child node of the anewarray node is a component class load from
+   // the original array object (array object to be copied). i.e. we are loading the component class from the original array class
+   // and pass the component class as 2nd child of anewarray. So the 2nd child must be evaluated into a reg.
+   bool dynamicArrayClone = (node->getOpCodeValue() == TR::anewarray) && (node->getSecondChild()->getSymbolReference()) &&
+         (node->getSecondChild()->getSymbolReference() == comp->getSymRefTab()->findArrayComponentTypeSymbolRef());
```

Comment:
this seems a little optimistic - shouldn't we say that if it isn't a loadaddr (or possibly aconst) then it is dynamic and assert that we have clazz if the array is not dynamic?

Reply:
Yes, added extra condition that checks for the class node op code for dynamic array, also assertion that rensure clazz cannot be null when not dynamic array.

Reply:
This still seems pretty special-purpose. Do we need to require that the component type child be either constant or in this restricted form? Or could we use something like the following?
```
node->getOpCodeValue() == TR::anewarray && clazz == NULL
```
I think `canAllocateInline()` could set `clazz = NULL` whenever the child isn't `loadaddr`.

Reply:
I revised the logic for evaluating the `isDynamicArray` boolean. Now we only initialize the boolean once in the top level NewEvaluator using the return values from `canAllocateInline()`. This way checking the `ArrayComponentType` symbol occurs only once in `canAllocateInline()`. And the NewEvaluator passes this flag down to helper routines which only need to check the flag to do the right thing. Currently the boolean is named as array-specific since only dynamic arrays are supported. When we implement dynamic object we can rename it properly -- just `isDynamicAllocation`.